src/core/raster/qgsrastercontourrenderer.cpp

```diff
@@ -0,0 +1,226 @@
+        r->setContourSymbol( static_cast<QgsLineSymbol *>( symbol ) );
+    }
+    if ( symbolMap.contains( QStringLiteral( "index-contour" ) ) )
+    {
+      QgsSymbol *symbol = symbolMap.take( QStringLiteral( "index-contour" ) );
+      if ( symbol->type() == QgsSymbol::Line )
+        r->setContourIndexSymbol( static_cast<QgsLineSymbol *>( symbol ) );
+    }
+  }
+  return r;
+}
+
+void QgsRasterContourRenderer::writeXml( QDomDocument &doc, QDomElement &parentElem ) const
+{
+  if ( parentElem.isNull() )
+  {
+    return;
+  }
+
+  QDomElement rasterRendererElem = doc.createElement( QStringLiteral( "rasterrenderer" ) );
+  _writeXml( doc, rasterRendererElem );
+
+  rasterRendererElem.setAttribute( QStringLiteral( "band" ), mInputBand );
+  rasterRendererElem.setAttribute( QStringLiteral( "contour-interval" ), mContourInterval );
+  rasterRendererElem.setAttribute( QStringLiteral( "contour-index-interval" ), mContourIndexInterval );
+  rasterRendererElem.setAttribute( QStringLiteral( "downscale" ), mDownscale );
+
+  QgsSymbolMap symbols;
+  symbols[QStringLiteral( "contour" )] = mContourSymbol.get();
+  if ( mContourIndexSymbol )
+    symbols[QStringLiteral( "index-contour" )] = mContourIndexSymbol.get();
+  QDomElement symbolsElem = QgsSymbolLayerUtils::saveSymbols( symbols, QStringLiteral( "symbols" ), doc, QgsReadWriteContext() );
+  rasterRendererElem.appendChild( symbolsElem );
+
+  parentElem.appendChild( rasterRendererElem );
+}
+
+struct ContourWriterData
+{
+  QPainter *painter;
+  double scaleX, scaleY;
+  QgsLineSymbol *symbol;
+  QgsLineSymbol *indexSymbol;
+  double indexInterval;
+  QgsRenderContext *context;
+};
+
+CPLErr _rasterContourWriter( double dfLevel, int nPoints, double *padfX, double *padfY, void *ptr )
+{
+  Q_UNUSED( dfLevel )
+  ContourWriterData *crData = static_cast<ContourWriterData *>( ptr );
+  QPolygonF polygon( nPoints );
+  QPointF *d = polygon.data();
+  for ( int i = 0; i < nPoints; ++i )
+  {
+    d[i] = QPointF( padfX[i] * crData->scaleX, padfY[i] * crData->scaleY );
+  }
+
+  if ( crData->indexSymbol && !qgsDoubleNear( crData->indexInterval, 0 ) && qgsDoubleNear( fmod( dfLevel, crData->indexInterval ), 0 ) )
+    crData->indexSymbol->renderPolyline( polygon, nullptr, *crData->context );
+  else
+    crData->symbol->renderPolyline( polygon, nullptr, *crData->context );
+  return CE_None;
+}
+
+QgsRasterBlock *QgsRasterContourRenderer::block( int bandNo, const QgsRectangle &extent, int width, int height, QgsRasterBlockFeedback *feedback )
+{
+  Q_UNUSED( bandNo )
+
+  std::unique_ptr< QgsRasterBlock > outputBlock( new QgsRasterBlock() );
+  if ( !mInput || !mContourSymbol )
+  {
+    return outputBlock.release();
+  }
+
+  int inputWidth = static_cast<int>( round( width / mDownscale ) );
+  int inputHeight = static_cast<int>( round( height / mDownscale ) );
+
+  std::shared_ptr< QgsRasterBlock > inputBlock( mInput->block( mInputBand, extent, inputWidth, inputHeight, feedback ) );
```

Comment:
unique_ptr would be more appropriate

Reply:
true, not sure how that shared_ptr got in there ... :-)